Services/Administration/OJS.Services.Administration.Business/Implementations/SubmissionsBusinessService.cs

```diff
@@ -156,6 +161,16 @@ public async Task<ServiceResult> Retest(Submission submission)
             var submissionProblemId = submission.ProblemId!.Value;
             var submissionParticipantId = submission.ParticipantId!.Value;
 
+            var problem = await this.problemsDataService.GetWithProblemGroupCheckerAndTestsById(submissionProblemId);
+            if (problem == null)
+            {
+                throw new BusinessServiceException("Problem cannot be null");
+            }
+
+            var submissionType = problem.SubmissionTypesInProblems
+                .First(st => st.SubmissionTypeId == submission.SubmissionTypeId)
+                .SubmissionType;
```

Comment:
Why do you get the Problem again? Everything you need (the execution context) is already in the submission.
We need to store the json of `submission.Map<SubmissionServiceModel>()`

Reply:
Reworked.